### PR TITLE
Php8 support

### LINKED ISSUE #13: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by marcelthole**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | yes

### Description
- fixes #13 
- Removes the `Null` Logger. This is a BC break

### CODE REVIEW

File: src/Writer/Null.php [hunk 1]
```diff
@@ -1,34 +0,0 @@
...
- * Stub class for backwards compatibility.
- *
- * Since PHP 7 adds "null" as a reserved keyword, we can no longer have a class
- * named that and retain PHP 7 compatibility. The original class has been
- * renamed to "Noop", and this class is now an extension of it. It raises an
- * E_USER_DEPRECATED to warn users to migrate.
- *
- * @deprecated
- */
-class Null extends Noop
```
Comment (rieschl) [inline src/Writer/Null.php:21]: I think deleting the class introduces a bc break. You shouldn't do that, unless the maintainers want to release a new major version (which the probably don't want to).

File: test/ProcessorPluginManagerFactoryTest.php [hunk 2]
```diff
@@ -27,7 +30,10 @@ public function testFactoryReturnsPluginManager()
...
         if (method_exists($processors, 'configure')) {
             // laminas-servicemanager v3
-            $this->assertAttributeSame($container, 'creationContext', $processors);
+            $creationContext = \Closure::bind(function () {
+                return $this->creationContext;
+            }, $processors, ProcessorPluginManager::class)();
+            $this->assertSame($container, $creationContext);
         } else {
             // laminas-servicemanager v2
             $this->assertSame($container, $processors->getServiceLocator());
```
Comment (rieschl) [inline test/ProcessorPluginManagerFactoryTest.php]: You can drop the whole `else` branch as `ServiceManager` v2 was also removed

File: test/FilterPluginManagerFactoryTest.php [hunk 3]
```diff
@@ -27,7 +30,10 @@ public function testFactoryReturnsPluginManager()
 
         if (method_exists($filters, 'configure')) {
             // laminas-servicemanager v3
-            $this->assertAttributeSame($container, 'creationContext', $filters);
+            $creationContext = \Closure::bind(function () {
+                return $this->creationContext;
+            }, $filters, FilterPluginManager::class)();
+            $this->assertSame($container, $creationContext);
         } else {
```
Comment (rieschl) [inline test/FilterPluginManagerFactoryTest.php]: You can drop the whole `else` branch as `ServiceManager` v2 was also removed

File: test/FilterPluginManagerFactoryTest.php [hunk 4]
```diff
@@ -54,7 +60,7 @@ public function testFactoryConfiguresPluginManagerUnderContainerInterop()
     /**
      * @depends testFactoryReturnsPluginManager
      */
-    public function testFactoryConfiguresPluginManagerUnderServiceManagerV2()
+    public function testFactoryConfiguresPluginManagerUnderServiceManagerV2(): void
```
Comment (rieschl) [inline test/FilterPluginManagerFactoryTest.php]: You can drop the whole test as `ServiceManager` v2 was also removed

File: test/FormatterPluginManagerFactoryTest.php [hunk 5]
```diff
@@ -27,7 +30,10 @@ public function testFactoryReturnsPluginManager()
...
         if (method_exists($formatters, 'configure')) {
             // laminas-servicemanager v3
-            $this->assertAttributeSame($container, 'creationContext', $formatters);
+            $creationContext = \Closure::bind(function () {
+                return $this->creationContext;
+            }, $formatters, FormatterPluginManager::class)();
+            $this->assertSame($container, $creationContext);
         } else {
             // laminas-servicemanager v2
             $this->assertSame($container, $formatters->getServiceLocator());
```
Comment (rieschl) [inline test/FormatterPluginManagerFactoryTest.php]: You can drop the `else` branch as `ServiceManager` v2 was also removed

File: test/FormatterPluginManagerFactoryTest.php [hunk 6]
```diff
@@ -54,7 +60,7 @@ public function testFactoryConfiguresPluginManagerUnderContainerInterop()
     /**
      * @depends testFactoryReturnsPluginManager
      */
-    public function testFactoryConfiguresPluginManagerUnderServiceManagerV2()
+    public function testFactoryConfiguresPluginManagerUnderServiceManagerV2(): void
```
Comment (rieschl) [inline test/FormatterPluginManagerFactoryTest.php]: You can drop the test as `ServiceManager` v2 was also removed

**Review by rieschl — CHANGES_REQUESTED** [review]
Nice PR! You still have some `if/else` conditions for `ServiceManager` `v2/v3` which you can drop because you also removed v2 from composer.

File: src/Writer/Null.php (hunk 1 above)
Comment (marcelthole) [reply src/Writer/Null.php:21]: In PHP 8  `Null` is a reserved keyword and produce an fatal error. https://3v4l.org/lKAQ4
Since its deprecated already i deleted this class. So yeah, yo are right. This is a BC break. I will update the PR description.

Comment (rieschl) [reply src/Writer/Null.php:21]: Yes, it's already reserved since PHP7. But as long you don't use it, there is no problem.
But I don't know what's the bc break policy here, because if you drop PHP 5, the class will be useless anyway. maybe @boesing or @Xerkus know more?

Comment (boesing) [reply src/Writer/Null.php:21]: As the class is not instantiable in any of those versions we support within `composer.json`, deleting this class is **not** a BC break as it cannot be used in either of these versions anyway.

File: composer.json [hunk 7]
```diff
@@ -30,8 +30,8 @@
         }
     },
     "require": {
-        "php": "^5.6 || ^7.0",
-        "laminas/laminas-servicemanager": "^2.7.5 || ^3.0.3",
+        "php": "^7.3 || ~8.0.0",
+        "laminas/laminas-servicemanager": "^3.3.0",
         "laminas/laminas-stdlib": "^2.7 || ^3.0",
```
Comment (boesing) [inline composer.json]: feel free to drop stdlib support for <3.0 aswell.

File: src/Writer/Stream.php [hunk 8]
```diff
@@ -76,7 +76,7 @@ public function __construct($streamOrUrl, $mode = null, $logSeparator = null, $f
             }
 
             $this->stream = $streamOrUrl;
-        } else {
+        } elseif (is_string($streamOrUrl)) {
```
Comment (boesing) [inline src/Writer/Stream.php]: is this related to PHP 8 or just a random bugfix?
If the latter, please create a dedicated PR regarding this bug.

File: composer.json [hunk 9]
```diff
@@ -30,8 +30,8 @@
         }
     },
     "require": {
-        "php": "^5.6 || ^7.0",
-        "laminas/laminas-servicemanager": "^2.7.5 || ^3.0.3",
+        "php": "^7.3 || ~8.0.0",
+        "laminas/laminas-servicemanager": "^3.3.0",
```
Comment (boesing) [inline composer.json:34]: please drop all SM v2 related code in here aswell 👍

**Review by boesing — CHANGES_REQUESTED** [review]

File: src/Writer/Stream.php (hunk 8 above)
Comment (marcelthole) [reply src/Writer/Stream.php]: its related to PHP 8 because in the testcase `xml_parser_create` is used. 
The function will return in PHP8 an XmlParser object instead of a resource. See https://3v4l.org/tc867 

To prevent type errors i added the `is_string` check. If the input is an Object you will get the `\Laminas\Log\Exception\InvalidArgumentException` instead of an typerror.
> Fatal error: Uncaught TypeError: fopen(): Argument 1 ($filename) must be of type string, XmlParser given

Any other ideas to handle this case? I think the library InvalidArgumentException is better than a TypeError?

**Comment by boesing** [comment]
Thanks, @marcelthole!
